### ISSUE TITLE
Memory usage in XGBoost training

### ISSUE THREAD

**Issue by dthiagarajan** [dmlc/xgboost#9254]
I'm trying to understand roughly where the bottlenecks for XGBoost memory usage will come up during training, assuming that I am
* Using the `hist` algorithm for trainig
* Loading all training and validation data into memory
* Training distributed using MPI + `rabit`

My understanding is that the main bottlenecks would be
* Before training, when loading all the data into memory
* During training, when nodes are added, due to computing the gradient histograms

For the latter, are the gradient histograms retained in memory throughout the whole training job? Or are they computed, temporarily stored to figure out how to split the tree, and then deleted?

I've noticed in some of my training jobs that the memory usage steadily increases over time, but I would assume that if the gradient histograms are only stored temporarily, this would not be the case. Can anyone clarify if this understanding is correct or not?

**Comment by trivialfis** [comment]
For CPU, it's kept for caching during each boosting iteration, and freed after a tree is built.

**Comment by dthiagarajan** [comment]
Got it. Would there be anything else that would be accumulating additional memory as each tree is built, besides the model itself.

**Comment by trivialfis** [comment]
I can't think of anything on top of my mind. Please note that the model can be non-trivial if you grow deep trees. Consider a tree with depth $d$ and dense nodes (full tree), the number of nodes is $2^d - 1$. Each node consumes at least 180 bytes due to storage of index, split value, hessian value, etc.

**Comment by dthiagarajan** [comment]
I see, that makes sense. What do you mean by the `index` that is stored at each node?

I'm also not clear on why each node requires it's own histogram. Can't the histogram be computed for the whole tree, and then based on the split value at each node, you can determine what the gradients are from the histogram associated with the tree? I saw [this comment](https://discuss.xgboost.ai/t/dask-is-distributed-training-globally-data-parallel-and-locally-feature-parallel/1929/11) on the forums that implied otherwise, but I don't quite get why given the following for a node in the tree 
- the histogram for the tree
- the split value (or the bounds for the feature)
you couldn't determine the gradient sums for the data associated with that node.

**Comment by trivialfis** [comment]
I will look into it a bit later. In general, if there's no memory leak, it's considered an optimization request and we might not be able to prioritize it immediately.

**Comment by dthiagarajan** [comment]
Thanks! Is there a chance it's due to something that's enabled when I set `verbosity=3`? I found that when I don't use verbosity, I don't see any issues (no increase in memory usage aside from creating the training QuantileDMatrix).

**Comment by weitian** [comment]
> It can not. Each node is associated with a partition of samples, and the histogram is computed with this partition. I recommend looking into decision trees in general.

Just want to confirm that, for distributed training,  it is by design that each node can have different histogram of samples?
We use QuntileDmatrix + dask_cuda.LocalCUDACluster with an 8 GPUs host for pairwise training.
On each Dask node, we construct the training data as QuntileDmatrix with around 500452 groups and 31885080 samples with 404 features.

**Comment by dthiagarajan** [comment]
I think I'm still not following the part about the histogram being recomputed for each node. If we had the histogram for the whole tree, and each node has an associated feature index and split value for that feature, couldn't we compute the histogram for that node by simply looking up the part of overall histogram that's associated with the feature index + split value?

And re: profilers, do you have any suggestions for what I could use for profiling?

**Comment by trivialfis** [comment]
> And the training hung sometimes with v1.7.4, but seemed working on v1.7.5.

We have some bug fixes.

On the tree node part, let me try explain the decision tree-building process here briefly. When a node is split, it splits on a feature based on a condition like $f_1 < 3.14$. The split essentially partitions the samples into two parts: samples with $f_1 < 3.14$ and samples with $f_1 >= 3.14$. Starting with the root node before it's split, we compute a histogram from all samples. Once the root node is split into two children, each owns a partition of the samples. The left node owns $f_1 < 3.14$ while the right one owns $f_1 >= 3.14$. You see, the children cannot directly reuse the histogram for the root node as they own different samples. We need to recompute the histogram from each partition by calculating the histogram for the left node from the partition where samples have $f_1 < 3.14$, and similarly, for the right node. There's no tree histogram since we don't have a tree.

> Just want to confirm that, for distributed training, it is by design that each node can have different histogram of samples?

I am trying to understand what this means. All workers share the same tree structure. Each worker builds its local histogram for a tree node, then allreduce them into a single histogram.

**Comment by dthiagarajan** [comment]
I'm still a bit confused about why it can't reuse the parent's histogram, maybe let me write out with an example to clarify. Is it because the split value is not guaranteed to be a bin edge?

For example, if I had a histogram like
```
bin_edges = [0, 1, 2, 3, 4]
counts = [3, 4, 5, 2]
```
If the split value was 2, then I could immediately compute the children's histograms as
```
# <= 2
bin_edges = [0, 1, 2]  # this could be a view of the original bin_edges
counts = [3, 4]  # this could be a view of the original counts
# > 2
bin_edges = [2, 3, 4]   # this could be a view of the original bin_edges
counts = [5, 2]   # this could be a view of the original counts
```

But if the split value was something like 2.5, and there were examples where the feature value was e.g. `2.1, 2.4, 2.45, 2.9`, then I could see that requiring to recompute the histograms.

**Comment by trivialfis** [comment]
The histogram is a histogram of gradient values used to approximate the gain function.

**Comment by adfea9c0** [comment]
Your observation is only true for the histogram of the feature that is being split on. If we split on feature `x1`, then the new histograms for feature `x2` are not trivially derivable from the parent's histogram.

**Comment by dthiagarajan** [comment]
Oh I see.. so to clarify:
* For the parent node, we compute histograms for each feature, where the histogram maps from bin edges for that feature to gradient values
* We decide which feature and what value for that feature to split on. From this, we can only get the histogram of the feature that is split on directly from the parent node, but otherwise, we need to create a new histogram for every other feature, for each child.

Is that right?
